aiida/restapi/resources.py

```diff
@@ -204,6 +204,122 @@ def get(self, id=None, page=None):  # pylint: disable=redefined-builtin,invalid-
         return self.utils.build_response(status=200, headers=headers, data=data)
 
 
+class QueryBuilder(BaseResource):
+    """
+    "Abstract" QueryBuilder REST API resource.
+
+    It is abstract in the sense that it supports POST requests taking in JSON QueryBuilder query help
+    objects and returning the QueryBuilder result accordingly.
+    """
+    from aiida.restapi.translator.nodes.node import NodeTranslator
+
+    _translator_class = NodeTranslator
+    _parse_pk_uuid = 'uuid'  # Parse a uuid pattern in the URL path (not a pk)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        # HTTP Request method decorators
+        if 'get_decorators' in kwargs and isinstance(kwargs['get_decorators'], (tuple, list, set)):
+            self.method_decorators.update({'post': list(kwargs['get_decorators'])})
+
+    def get(self):  # pylint: disable=arguments-differ
+        """Static return to state information about this endpoint."""
+        data = {
+            'message': (
+                'Use HTTP POST requests to use the AiiDA QueryBuilder. '
+                'POST JSON data, which MUST be a valid QueryBuilder queryhelp JSON object. '
+                'See the documentation at https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/'
+                'database.html?highlight=QueryBuilder#the-queryhelp for more information.'
+            ),
+        }
+
+        headers = self.utils.build_headers(url=request.url, total_count=1)
+        return self.utils.build_response(
+            status=200,
+            headers=headers,
+            data={
+                'method': request.method,
+                'url': unquote(request.url),
+                'url_root': unquote(request.url_root),
+                'path': unquote(request.path),
+                'query_string': request.query_string.decode('utf-8'),
+                'resource_type': 'Query',
+                'data': data,
+            },
+        )
+
+    def post(self):
+        """
+        POST method to pass query help JSON.
+
+        This assumes the posted data is a complete and valid QueryBuilder queryhelp JSON object.
+        We could create a RequestParser that completely redefines all valid keys in the queryhelp,
+        but for now, it's much easier to simply pass-through the posted queryhelp and make sure
+        it can be handled by the built-in NodeTranslator.
```

Comment:
Just to understand - is this because of limitations of this endpoint with respect to what the QueryBuilder can do?

In general it certainly would seem preferable if we don't have to add any extra translation step here.

Reply:
No, it was more to verify the POSTed JSON data.